### PR TITLE
add proper ipam resource implementation for EnsureCreated

### PR DESCRIPTION
**Pull request by xh3b4sd**
Towards Node Pools. Finally putting all the things together to make the `ipam` resource work in an abstracted way. 🎉

### CODE REVIEW

File: service/controller/clusterapi/v29/resource/ipam/create_test.go [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+	"context"
+	"net"
+	"strconv"
+	"testing"
+
+	"github.com/giantswarm/micrologger/microloggertest"
+	"sigs.k8s.io/cluster-api/pkg/apis/cluster/v1alpha1"
+)
+
+func Test_SubnetAllocator(t *testing.T) {
```
Comment (xh3b4sd) [inline service/controller/clusterapi/v29/resource/ipam/create_test.go]: The resource implementation as well as the unit tests are ported from the `network` package. I am going to drop it as the resource here replaces it now properly with additional capabilities of working with different types. 

Note that the tests are different in error handling as the matching for the allocated subnets is handled differently. It didn't appear to make sense to test this. When you think differently, this test should be extensible to do it rather easily.

File: service/controller/clusterapi/v29/resource/ipam/create_test.go [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+		publicSubnetMaskBits    int
+	}{
+		{
+			name:         "case 0 allocate first subnet",
+			object:       &v1alpha1.Cluster{},
+			errorMatcher: nil,
+
+			checker:   NewTestChecker(true),
+			collector: NewTestCollector([]net.IPNet{}),
+			persister: NewTestPersister(mustParseCIDR("10.100.0.0/24")),
```
Comment (xh3b4sd) [inline service/controller/clusterapi/v29/resource/ipam/create_test.go]: The parameters to the test constructors could take a config object with named properties to make it clearer what these parameters are about. That wasn't too important to me. Let me know when you would like to see this. Then we can just extend the code accordingly.

File: service/controller/clusterapi/v29/resource/ipam/test_persister.go [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+	p := &TestPersister{
+		subnet: subnet,
+	}
+
+	return p
+}
+
+func (p *TestPersister) Persist(ctx context.Context, subnet net.IPNet, namespace string, name string) error {
+	if !reflect.DeepEqual(subnet, p.subnet) {
+		return microerror.Mask(invalidConfigError)
```
Comment (xh3b4sd) [inline service/controller/clusterapi/v29/resource/ipam/test_persister.go]: I was lazy and only reused the error the package already provided. Let me know when you want to see that differently.

File: service/controller/clusterapi/v29/resource/ipam/create.go [hunk 4]
```diff
@@ -2,10 +2,67 @@ package ipam
...
+		}
+
+		r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("found allocated subnets %#q", allocatedSubnets))
+	}
+
+	var freeSubnet net.IPNet
+	{
+		r.logger.LogCtx(ctx, "level", "debug", "message", "finding free subnet")
+
+		freeSubnet, err = ipam.Free(r.networkRange, r.allocatedSubnetMask, allocatedSubnets)
```
Comment (rossf7) [inline service/controller/clusterapi/v29/resource/ipam/create.go]: Really nice the package design makes the code very readable.

File: service/controller/clusterapi/v29/resource/ipam/create_test.go (hunk 1 above)
Comment (rossf7) [reply service/controller/clusterapi/v29/resource/ipam/create_test.go]: Not deep into this topic but that makes sense to me.

**Pull request merged**